File: skeleton/intelligence/cognition.py
"""Cognition — the empire's belief engine.

Port of gameforge-rs ``cognition.rs``: beliefs are claims with confidence;
evidence moves confidence in log-odds space so no single witness can flip
the mind, but many independent witnesses can. Claims sharing a predicate
with opposite polarity and high confidence are flagged as **schisms**.

Pure domain — sync Python (threading.Lock), optional injected clock for tests.
No FastAPI / lifespan.
"""
from __future__ import annotations

import math
import threading
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence

WITNESS_STEP: float = 1.2
CONVICTION: float = 0.3
DECAY_PER_DAY: float = 0.05
LEDGER_CAP: int = 8192


@dataclass
class Evidence:
    witness: str
    polarity: bool  # True = supports, False = refutes
    weight: float  # 0.0..=1.0 witness reliability
    at: float  # unix seconds


@dataclass
class Belief:
    id: str
    predicate: str
    polarity: bool
    lodds: float = 0.0
    evidence: List[Evidence] = field(default_factory=list)
    created: float = 0.0
    touched: float = 0.0

    def confidence(self) -> float:
        return 1.0 / (1.0 + math.exp(-self.lodds))


@dataclass
class Schism:
    predicate: str
    supporting: List[str]  # belief ids
    refuting: List[str]
    detected: float  # unix seconds


class Cognition:
    """Hold beliefs, ingest witness evidence, detect and resolve schisms."""

    def __init__(self, *, clock: Optional[Callable[[], float]] = None) -> None:
        self._clock: Callable[[], float] = clock if clock is not None else time.time
        self._beliefs: Dict[str, Belief] = {}
        self._schisms: List[Schism] = []
        self._lock = threading.Lock()

    def _now(self) -> float:
        return float(self._clock())

    def hold(self, predicate: str, polarity: bool) -> str:
        """Hold a new belief (or locate the existing one on predicate+polarity)."""
        with self._lock:
            for b in self._beliefs.values():
                if b.predicate == predicate and b.polarity == polarity:
                    return b.id
            if len(self._beliefs) >= LEDGER_CAP:
                by_conviction = sorted(
                    ((bid, abs(b.lodds)) for bid, b in self._beliefs.items()),
                    key=lambda kv: kv[1],
                )
                for bid, _ in by_conviction[: LEDGER_CAP // 4]:
                    self._beliefs.pop(bid, None)
            now = self._now()
            bid = str(uuid.uuid4())
            self._beliefs[bid] = Belief(
                id=bid,
                predicate=predicate,
                polarity=polarity,
                lodds=0.0,
                evidence=[],
                created=now,
                touched=now,
            )
            return bid

    def testify(
        self,
        belief_id: str,
        witness: str,
        supports: bool,
        weight: float,
    ) -> Optional[float]:
        """Add witness evidence; updates log-odds and re-scans for schisms."""
        weight = max(0.0, min(1.0, float(weight)))
        with self._lock:
            b = self._beliefs.get(belief_id)
            if b is None:
                return None
            now = self._now()
            days = (now - b.touched) / 86400.0
            if days > 0.0:
                b.lodds *= max(0.0, 1.0 - DECAY_PER_DAY * days)
            prior = sum(1 for e in b.evidence if e.witness == witness)
            voice = weight / (1.0 + prior)
            step = WITNESS_STEP * voice * (1.0 if supports else -1.0)
            b.lodds = max(-8.0, min(8.0, b.lodds + step))
            b.evidence.append(
                Evidence(witness=witness, polarity=supports, weight=weight, at=now)
            )
            b.touched = now
            predicate = b.predicate
            confidence = b.confidence()
        self.scan_schism(predicate)
        return confidence

    def scan_schism(self, predicate: str) -> None:
        """Contradiction scan on one predicate. Schism needs conviction on both sides."""
        with self._lock:
            supporting = [
                b.id
                for b in self._beliefs.values()
                if b.predicate == predicate
                and b.polarity
                and b.confidence() >= 0.5 + CONVICTION
            ]
            refuting = [
                b.id
                for b in self._beliefs.values()
                if b.predicate == predicate
                and not b.polarity
                and b.confidence() >= 0.5 + CONVICTION
            ]
            if not supporting or not refuting:
                return
            if any(s.predicate == predicate for s in self._schisms):
                return
            self._schisms.append(
                Schism(
                    predicate=predicate,
                    supporting=list(supporting),
                    refuting=list(refuting),
                    detected=self._now(),
                )
            )
            if len(self._schisms) > 256:
                n = len(self._schisms) - 256
                del self._schisms[:n]

    def resolve_schism(self, predicate: str, winning_polarity: bool) -> bool:
        """Court decision: clear losing polarity lodds and strike the schism."""
        with self._lock:
            idx = next(
                (i for i, s in enumerate(self._schisms) if s.predicate == predicate),
                None,
            )
            if idx is None:
                return False
            self._schisms.pop(idx)
            for b in self._beliefs.values():
                if b.predicate == predicate and b.polarity != winning_polarity:
                    b.lodds = 0.0
            return True

    def belief(self, belief_id: str) -> Optional[Belief]:
        with self._lock:
            b = self._beliefs.get(belief_id)
            if b is None:
                return None
            return Belief(
                id=b.id,
                predicate=b.predicate,
                polarity=b.polarity,
                lodds=b.lodds,
                evidence=list(b.evidence),
                created=b.created,
                touched=b.touched,
            )

    def schisms(self) -> List[Schism]:
        with self._lock:
            return [
                Schism(
                    predicate=s.predicate,
                    supporting=list(s.supporting),
                    refuting=list(s.refuting),
                    detected=s.detected,
                )
                for s in self._schisms
            ]

    def stats(self) -> Dict[str, Any]:
        with self._lock:
            convinced = sum(1 for b in self._beliefs.values() if abs(b.lodds) > WITNESS_STEP)
            return {
                "beliefs": len(self._beliefs),
                "convinced": convinced,
                "open_schisms": len(self._schisms),
            }

    def assert_plan_claims(
        self,
        claims: Sequence[Mapping[str, Any]],
    ) -> List[Schism]:
        """Hold + testify each plan claim (witness=plan, weight=1); return open schisms.

        For each ``{predicate, polarity}``, holds the belief and testifies once with
        witness ``"plan"`` / weight 1.0 supporting the claim. Returns currently open
        schisms whose predicate appears in the claim set (newly opened or pre-existing).
        """
        predicates: List[str] = []
        for raw in claims:
            pred = str(raw.get("predicate") or "").strip()
            if not pred:
                continue
            polarity = bool(raw.get("polarity", True))
            predicates.append(pred)
            bid = self.hold(pred, polarity)
            self.testify(bid, "plan", True, 1.0)
        wanted = set(predicates)
        return [s for s in self.schisms() if s.predicate in wanted]

File: satellites/gameforge-rs/crates/gf-gameforge/src/cognition.rs
use std::collections::HashMap;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Evidence { pub witness: String, pub polarity: bool, pub weight: f64, pub at: DateTime<Utc> }
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Belief {
    pub id: String, pub predicate: String, pub polarity: bool, pub lodds: f64,
    pub evidence: Vec<Evidence>, pub created: DateTime<Utc>, pub touched: DateTime<Utc>,
}
impl Belief { pub fn confidence(&self) -> f64 { 1.0 / (1.0 + (-self.lodds).exp()) } }
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Schism { pub predicate: String, pub supporting: Vec<String>, pub refuting: Vec<String>, pub detected: DateTime<Utc> }
const WITNESS_STEP: f64 = 1.2;
const CONVICTION: f64 = 0.3;
const DECAY_PER_DAY: f64 = 0.05;
const LEDGER_CAP: usize = 8192;

pub struct Cognition { beliefs: RwLock<HashMap<String, Belief>>, schisms: RwLock<Vec<Schism>> }
impl Default for Cognition { fn default() -> Self { Self::new() } }
impl Cognition {
    pub fn new() -> Self { Self { beliefs: RwLock::new(HashMap::new()), schisms: RwLock::new(Vec::new()) } }
    pub async fn hold(&self, predicate: &str, polarity: bool) -> String {
        let mut beliefs = self.beliefs.write().await;
        for b in beliefs.values() { if b.predicate == predicate && b.polarity == polarity { return b.id.clone(); } }
        if beliefs.len() >= LEDGER_CAP {
            let mut by_conviction: Vec<(String, f64)> = beliefs.iter().map(|(id,b)|(id.clone(), b.lodds.abs())).collect();
            by_conviction.sort_by(|a,b| a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal));
            for (id,_) in by_conviction.into_iter().take(LEDGER_CAP / 4) { beliefs.remove(&id); }
        }
        let id = Uuid::new_v4().to_string();
        beliefs.insert(id.clone(), Belief { id: id.clone(), predicate: predicate.to_string(), polarity, lodds: 0.0, evidence: Vec::new(), created: Utc::now(), touched: Utc::now() });
        id
    }
    pub async fn testify(&self, belief_id: &str, witness: &str, supports: bool, weight: f64) -> Option<f64> {
        let weight = weight.clamp(0.0,1.0);
        let mut beliefs = self.beliefs.write().await;
        let now = Utc::now();
        let b = beliefs.get_mut(belief_id)?;
        let days = (now - b.touched).num_seconds() as f64 / 86400.0;
        if days > 0.0 { b.lodds *= (1.0 - DECAY_PER_DAY * days).max(0.0); }
        let prior = b.evidence.iter().filter(|e| e.witness == witness).count() as f64;
        let voice = weight / (1.0 + prior);
        b.lodds = (b.lodds + WITNESS_STEP * voice * if supports {1.0} else {-1.0}).clamp(-8.0,8.0);
        b.evidence.push(Evidence { witness: witness.to_string(), polarity: supports, weight, at: now });
        b.touched = now;
        let predicate = b.predicate.clone();
        let confidence = b.confidence();
        drop(beliefs);
        self.scan_schism(&predicate).await;
        Some(confidence)
    }
    async fn scan_schism(&self, predicate: &str) {
        let beliefs = self.beliefs.read().await;
        let supporting: Vec<String> = beliefs.values().filter(|b| b.predicate == predicate && b.polarity && b.confidence() >= 0.5 + CONVICTION).map(|b|b.id.clone()).collect();
        let refuting: Vec<String> = beliefs.values().filter(|b| b.predicate == predicate && !b.polarity && b.confidence() >= 0.5 + CONVICTION).map(|b|b.id.clone()).collect();
        drop(beliefs);
        if supporting.is_empty() || refuting.is_empty() { return; }
        let mut schisms = self.schisms.write().await;
        if !schisms.iter().any(|s| s.predicate == predicate) {
            schisms.push(Schism { predicate: predicate.to_string(), supporting, refuting, detected: Utc::now() });
            if schisms.len() > 256 { let n = schisms.len()-256; schisms.drain(0..n); }
        }
    }
    pub async fn resolve_schism(&self, predicate: &str, winning_polarity: bool) -> bool {
        let mut schisms = self.schisms.write().await;
        let Some(idx) = schisms.iter().position(|s| s.predicate == predicate) else { return false };
        schisms.remove(idx); drop(schisms);
        let mut beliefs = self.beliefs.write().await;
        for b in beliefs.values_mut() { if b.predicate == predicate && b.polarity != winning_polarity { b.lodds = 0.0; } }
        true
    }
    pub async fn belief(&self, id:&str)->Option<Belief>{self.beliefs.read().await.get(id).cloned()}
    pub async fn schisms(&self)->Vec<Schism>{self.schisms.read().await.clone()}
    pub async fn stats(&self)->serde_json::Value{
        let beliefs=self.beliefs.read().await; let convinced=beliefs.values().filter(|b|b.lodds.abs()>WITNESS_STEP).count();
        serde_json::json!({"beliefs":beliefs.len(),"convinced":convinced,"open_schisms":self.schisms.read().await.len()})
    }
}
